File: 14/python/main.py
from collections import defaultdict
def optimized(lines):
    template = lines[0].strip()
    rules = [x.strip().split(" -> ") for x in lines[2:]]
    rules = {x: y for [x,y] in rules}
    # Map from pairs -> pairs
    pairs = {}
    for i in range(len(template)-1):
        val = template[i:i+2]
        if val in pairs:
            pairs[val] += 1
        else:
            pairs[val] = 1

    last_pair = template[-2:]
    for _ in range(0, 40):
        new_pairs = defaultdict(lambda: 0)
        for x in pairs:
            if x in rules:
                new_pairs[x[0] + rules[x]] += pairs[x]
                new_pairs[rules[x] + x[1]] += pairs[x]
            else:
                new_pairs[x] += pairs[x]
        pairs = new_pairs
        if last_pair in rules:
            last_pair = rules[last_pair] + last_pair[1]

    final_count = defaultdict(lambda: 0)
    for x in pairs:
        final_count[x[0]] += pairs[x]
    final_count[last_pair[1]] += 1
    z = (sorted([(final_count[x], x) for x in final_count]))
    print(z[-1][0] - z[0][0]) # Could be off by one in either direction

with open('input.txt') as f:
    lines = f.readlines()
    optimized(lines)

File: 14/haskell/main.hs
import Data.Either
import Data.List
import Data.Map
import System.IO
import Text.ParserCombinators.Parsec

type BasePair = (Char, Char)

type PairCounter = Map BasePair Int

type RuleMapEntry = (BasePair, (BasePair, BasePair))

type RuleMap = Map BasePair (BasePair, BasePair)

data Day14Input = Day14Input String RuleMap deriving (Show)

stringToBasePairs :: String -> [BasePair]
stringToBasePairs (a : b : xs) = (a, b) : stringToBasePairs (b : xs)
stringToBasePairs _ = []

stringToPairCounter :: String -> PairCounter
stringToPairCounter input =
  fromList $
    Data.List.map (\x -> (head x, length x)) $
      group $ sort $ stringToBasePairs input

buildRuleMapEntry :: BasePair -> Char -> RuleMapEntry
buildRuleMapEntry (a, b) c = ((a, b), ((a, c), (c, b)))

day14 :: CharParser () Day14Input
day14 =
  Day14Input
    <$> template
    <*> (string "\n\n" >> rules)

template :: CharParser () String
template = many (noneOf "\n")

rules :: CharParser () RuleMap
rules = fromList <$> sepBy rule (char '\n')

rule :: CharParser () RuleMapEntry
rule = do
  a <- upper
  b <- upper
  string " -> "
  buildRuleMapEntry (a, b) <$> upper

incrementBy :: Int -> Maybe Int -> Maybe Int
incrementBy amount curr = case curr of
  Just y -> Just (amount + y)
  Nothing -> Just amount

applyRules :: PairCounter -> RuleMap -> PairCounter
applyRules pairs rules = Data.Map.foldrWithKey update empty pairs
  where
    update key x acc = case Data.Map.lookup key rules of
      Just (pair1, pair2) -> Data.Map.alter (incrementBy x) pair1 (Data.Map.alter (incrementBy x) pair2 acc)
      Nothing -> Data.Map.alter (incrementBy x) key acc

-- Note that this scoring won't count the last charater. So it needs to be passed
-- in explicitly.
score :: Char -> PairCounter -> Int
score lastChar pairs = Data.List.last sorted - Data.List.head sorted
  where
    sorted = sort $elems $ Data.Map.foldrWithKey update (fromList [(lastChar, 1)]) pairs
    update (a, _) x acc = Data.Map.alter (incrementBy x) a acc

solution1 (Day14Input template rules) =
  score (Data.List.last template) $ Data.List.foldl (\acc _ -> applyRules acc rules) (stringToPairCounter template) [1 .. 10]

solution2 (Day14Input template rules) =
  score (Data.List.last template) $ Data.List.foldl (\acc _ -> applyRules acc rules) (stringToPairCounter template) [1 .. 40]

main = do
  handle <- openFile "input.txt" ReadMode
  contents <- hGetContents handle
  let parsed = fromRight undefined $ parse day14 "Invalid input" contents
  print (solution1 parsed)
  print (solution2 parsed)
  hClose handle
